File: connector/opcua/src/main/java/io/nexus/gateway/opcua/WriteHandler.java
// Copyright 2026 nexus-gateway contributors
// SPDX-License-Identifier: Apache-2.0

package io.nexus.gateway.opcua;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.Map;
import java.util.concurrent.*;
import java.util.stream.Collectors;

/**
 * Handles write commands on cmd.opcua.<connector_id> (ADR-0004).
 * Thread-safe: NATS dispatcher may deliver messages on multiple threads.
 */
public class WriteHandler {

    @FunctionalInterface
    public interface Responder {
        void reply(String replyTo, byte[] data) throws Exception;
    }

    private static final int DEDUP_MAX = 1000;
    private static final String IN_FLIGHT = "__in_flight__";
    private static final Logger log = LoggerFactory.getLogger(WriteHandler.class);

    private final Config cfg;
    private final OpcUaClientFacade client;
    private final Responder responder;
    private final Map<String, PointConfig> pointMap;
    private final ConcurrentHashMap<String, WriteReply> dedup = new ConcurrentHashMap<>();
    private final ExecutorService writeExecutor = Executors.newCachedThreadPool(r -> {
        Thread t = new Thread(r, "opcua-write");
        t.setDaemon(true);
        return t;
    });

    public WriteHandler(Config cfg, OpcUaClientFacade client, Responder responder) {
        this.cfg = cfg;
        this.client = client;
        this.responder = responder;
        this.pointMap = cfg.points().stream()
            .collect(Collectors.toMap(PointConfig::localId, p -> p));
    }

    public void handle(byte[] data, String replyTo) {
        ControlCommand cmd;
        try {
            cmd = ControlCommand.fromBytes(data);
        } catch (Exception ex) {
            log.warn("opcua: unparseable command: {}", ex.getMessage());
            send(replyTo, new WriteReply(false, "bad_request"));
            return;
        }

        // Idempotency: putIfAbsent with sentinel prevents concurrent double-writes (TOCTOU-safe).
        WriteReply sentinel = new WriteReply(false, IN_FLIGHT);
        WriteReply previous = dedup.putIfAbsent(cmd.controlId(), sentinel);
        if (previous != null) {
            if (!IN_FLIGHT.equals(previous.response())) {
                send(replyTo, previous);
            }
            // else: in-flight duplicate — drop silently; retry will arrive after result is stored
            return;
        }

        WriteReply result = execute(cmd);
        dedup.put(cmd.controlId(), result);
        evictIfNeeded();
        send(replyTo, result);
    }

    private WriteReply execute(ControlCommand cmd) {
        PointConfig pt = pointMap.get(cmd.localId());
        if (pt == null || !pt.writable()) {
            return new WriteReply(false, "not_writable");
        }

        long timeoutMs = (long) (cfg.writeTimeout() * 1000);
        Future<Void> future;
        try {
            future = writeExecutor.submit(() -> {
                if (pt.methodNodeId() != null) {
                    client.callMethod(cmd.localId(), pt.methodNodeId(), cmd.value());
                } else {
                    client.writeNode(cmd.localId(), cmd.value());
                }
                return null;
            });
        } catch (RejectedExecutionException ex) {
            log.warn("opcua: write executor unavailable for {} control_id={}", cmd.localId(), cmd.controlId());
            return new WriteReply(false, "device_error: executor_shutdown");
        }

        try {
            future.get(timeoutMs, TimeUnit.MILLISECONDS);
            log.info("opcua: wrote {}={} control_id={}", cmd.localId(), cmd.value(), cmd.controlId());
            return new WriteReply(true, "ok");
        } catch (TimeoutException ex) {
            future.cancel(true);
            log.warn("opcua: write timed out for {} control_id={}", cmd.localId(), cmd.controlId());
            return new WriteReply(false, "timeout");
        } catch (ExecutionException ex) {
            Throwable cause = ex.getCause() != null ? ex.getCause() : ex;
            log.warn("opcua: write device error for {}: {}", cmd.localId(), cause.getMessage());
            return new WriteReply(false, "device_error: " + cause.getMessage());
        } catch (InterruptedException ex) {
            Thread.currentThread().interrupt();
            return new WriteReply(false, "interrupted");
        }
    }

    private void evictIfNeeded() {
        // Skip IN_FLIGHT sentinels — evicting them breaks the TOCTOU-safe idempotency guarantee.
        while (dedup.size() > DEDUP_MAX) {
            boolean removed = false;
            for (Map.Entry<String, WriteReply> entry : dedup.entrySet()) {
                if (!IN_FLIGHT.equals(entry.getValue().response())) {
                    if (dedup.remove(entry.getKey(), entry.getValue())) {
                        removed = true;
                        break;
                    }
                }
            }
            if (!removed) break; // only sentinels remain; don't evict
        }
    }

    private void send(String replyTo, WriteReply reply) {
        if (replyTo == null || replyTo.isBlank()) return;
        try {
            responder.reply(replyTo, reply.encode());
        } catch (Exception ex) {
            log.warn("opcua: failed to send write reply: {}", ex.getMessage());
        }
    }
}

File: connector/bacnet/bacnet_connector/write_handler.py
# Copyright 2026 nexus-gateway contributors
# SPDX-License-Identifier: Apache-2.0

"""BACnet write handler: subscribe to cmd.<protocol>.<connector_id>, execute WriteProperty."""
from __future__ import annotations

import asyncio
import logging

from bacnet_connector.command import ControlCommand, WriteReply
from bacnet_connector.config import Config
from bacnet_connector.connector import BACnetClient

logger = logging.getLogger(__name__)

_DEDUP_MAX = 1000  # maximum cached control_ids kept in memory


class WriteHandler:
    """Handles write commands arriving on cmd.bacnet.<connector_id>.

    Designed for direct testing via handle() without a real NATS connection.
    Wire up the NATS subscription separately in main.py.
    """

    def __init__(self, cfg: Config, bacnet: BACnetClient) -> None:
        self._cfg = cfg
        self._bacnet = bacnet
        self._point_map = {pt.local_id: pt for pt in cfg.points}
        self._dedup: dict[str, WriteReply] = {}

    async def handle(self, msg: object) -> None:
        """Process one command message and reply via msg.respond()."""
        try:
            cmd = ControlCommand.from_bytes(msg.data)  # type: ignore[union-attr]
        except Exception as exc:
            logger.warning("bacnet: write handler received unparseable command: %s", exc)
            await msg.respond(WriteReply(success=False, response="bad_request").encode())  # type: ignore[union-attr]
            return

        # Idempotency: reserve sentinel before first await to prevent TOCTOU double-write.
        if cmd.control_id in self._dedup:
            cached = self._dedup[cmd.control_id]
            if cached is not None:
                await msg.respond(cached.encode())  # type: ignore[union-attr]
            return  # in-flight duplicate: drop silently

        self._dedup[cmd.control_id] = None  # type: ignore[assignment]  # reserve slot
        result = await self._execute(cmd)
        self._cache(cmd.control_id, result)
        await msg.respond(result.encode())  # type: ignore[union-attr]

    async def _execute(self, cmd: ControlCommand) -> WriteReply:
        pt = self._point_map.get(cmd.local_id)
        if pt is None or not pt.writable:
            return WriteReply(success=False, response="not_writable")

        priority = cmd.priority if cmd.priority > 0 else self._cfg.default_write_priority

        try:
            await asyncio.wait_for(
                self._bacnet.write_property(
                    self._cfg.bacnet_address,
                    self._cfg.bacnet_device_id,
                    cmd.local_id,
                    "presentValue",
                    cmd.value,
                    priority,
                ),
                timeout=self._cfg.write_timeout,
            )
            logger.info(
                "bacnet: wrote %s=%s priority=%d for control_id=%s",
                cmd.local_id, cmd.value, priority, cmd.control_id,
            )
            return WriteReply(success=True, response="ok")
        except asyncio.TimeoutError:
            logger.warning("bacnet: write timed out for %s control_id=%s", cmd.local_id, cmd.control_id)
            return WriteReply(success=False, response="timeout")
        except Exception as exc:
            logger.warning("bacnet: write device error for %s: %s", cmd.local_id, exc)
            return WriteReply(success=False, response=f"device_error: {exc}")

    def _cache(self, control_id: str, result: WriteReply) -> None:
        if len(self._dedup) >= _DEDUP_MAX:
            # Evict the oldest entry (insertion-ordered dict, Python 3.7+)
            del self._dedup[next(iter(self._dedup))]
        self._dedup[control_id] = result
